com.microsoft.mrtk.graphicstools.unity/Editor/Experimental/AmbientOcclusion/AmbientOcclusionTool.cs

```diff
@@ -0,0 +1,325 @@
+                    Handles.SphereHandleCap(0, _vertexs[i], Quaternion.identity, settings._coverageRadius * (1 - occlusion), EventType.Repaint);
+                }
+            }
+
+            if (settings._showHits)
+            {
+                for (int i = 0; i < _referenceVertexHits.Count; i++)
+                {
+                    Handles.color = settings._hitColor;
+                    Handles.SphereHandleCap(0, _referenceVertexHits[i].point, Quaternion.identity, settings._hitRadius, EventType.Repaint);
+                }
+            }
+
+            return;
+        }
+
+        /// <summary>
+        /// Get a random sample direction from a hemisphere who's base is 
+        /// perpendicular to the reference normal.
+        /// </summary>
+        /// <param name="normalizedReferenceNormal">Must be noramlized</param>
+        /// <returns></returns>
+        private Vector3 RandomSampleAboveHemisphere(Vector3 normalizedReferenceNormal)
+        {
+            float angle;
+            Vector3 randomAxis;
+            while (true)
+            {
+                // This method is appears to be more uniform than Random.onUnitSphere in testing...
+                Random.rotationUniform.ToAngleAxis(out angle, out randomAxis);
+                if (Vector3.Dot(randomAxis, normalizedReferenceNormal) > 0)
+                {
+                    return randomAxis;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Peform the ambient occlusion calculation
+        /// </summary>
+        internal void GatherSamples(GameObject go)
+        {
+            var meshFilter = go.GetComponent<MeshFilter>();
+            if (meshFilter == null)
+            {
+                Debug.LogWarning($"No mesh filter found for {go.name}! - skipping.", go);
+                return;
+            }
+
+            var mesh = DeepCopyMesh(meshFilter.sharedMesh);
+            mesh.name = $"{_magicPrefix} {go.name}";
+            meshFilter.mesh = mesh;
+
+            if (_bentNormalsAo == null || _bentNormalsAo.Length != mesh.vertexCount)
+            {
+                _bentNormalsAo = new Vector4[mesh.vertexCount];
+            }
+
+            var watch = Stopwatch.StartNew();
+            RaycastHit hit;
+
+            _vertexs = mesh.vertices;
+            _normals = mesh.normals;
+            _referenceVertexSamples.Clear();
+            _referenceVertexHits.Clear();
+
+            for (int vi = 0; vi < mesh.vertexCount; vi++)
+            {
+                // Do the work in world space
+                _vertexs[vi] = go.transform.TransformPoint(_vertexs[vi]);
+                _normals[vi] = go.transform.TransformVector(_normals[vi]);
+
+                _hitsInHemisphere.Clear();
+
+                Vector3 averageDir = Vector3.zero;
+
+                var origin = _vertexs[vi] + _normals[vi].normalized * settings._originNormalOffset;
+
+                Physics.queriesHitBackfaces = true;
```

Comment:
Wonder if this could be moved out of the top level for loop for more speed?

Reply:
Moved.